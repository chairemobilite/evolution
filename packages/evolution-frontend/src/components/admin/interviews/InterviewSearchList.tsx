/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';

import Interview from './InterviewSearchResult';
import { InterviewContext } from '../../../contexts/InterviewContext';
import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import InterviewsCreateLink from './InterviewsCreateLink';

type InterviewSearchListProps = {
    autoCreateIfNoData: boolean;
    queryData: URLSearchParams;
};

const InterviewSearchList: React.FunctionComponent<InterviewSearchListProps & WithTranslation> = (
    props: InterviewSearchListProps & WithTranslation
) => {
    const [data, setData] = React.useState<{ [key: string]: any }[]>([]);
    const [loading, setLoading] = React.useState(false);
    const { state, dispatch } = React.useContext(InterviewContext);
    const fetchIdRef = React.useRef(0);

    // Function to fetch data from the server, with paging and filtering
    const fetchData = React.useCallback(
        async (accessCode: string) => {
            // Give this fetch an ID
            const fetchId = ++fetchIdRef.current;

            // Set the loading state
            setLoading(true);

            // Make a query string from the filters

            try {
                const response = await fetch(`/api/interviews/interviewByCode?accessCode=${accessCode}`);
                if (fetchId !== fetchIdRef.current) {
                    // There was another query since, ignore
                    return;
                }
                if (response.status === 200) {
                    const jsonData = await response.json();
                    if (jsonData.interviews) {
                        setData(jsonData.interviews);
                        if (
                            !_isBlank(accessCode) &&
                            jsonData.interviews.length === 0 &&
                            props.autoCreateIfNoData &&
                            state.status !== 'creating'
                        ) {
                            // automatically create new interview with this access code
                            dispatch({
                                type: 'createNew',
                                username: `telephone_${(Math.ceil(Math.random() * 8999) + 1000).toString()}`,
                                queryData: props.queryData
                            });
                        }
                    }
                } else {
                    console.error('Invalid response code from server: ', response.status);
                }
            } catch (error) {
                console.error(`Error fetching user data from server: ${error}`);
                setData([]);
            } finally {
                if (fetchId === fetchIdRef.current) {
                    setLoading(false);
                }
            }
        },
        [state.status]
    );

    React.useEffect(() => {
        fetchData(state.responses.accessCode);
    }, [state.responses]);

    // TODO This is very specific to the interviewer use case, generalize it and use a function to generate username, maybe as prop
    return (
        <ul className="interview">
            {data.map((interview) => (
                <Interview key={interview.uuid} interview={interview} />
            ))}
            <InterviewsCreateLink queryData={props.queryData} />
        </ul>
    );
};

export default withTranslation(['admin', 'main'])(InterviewSearchList);
