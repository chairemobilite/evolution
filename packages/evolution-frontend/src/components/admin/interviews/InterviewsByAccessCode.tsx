/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { useTranslation } from 'react-i18next';
import Loadable from 'react-loadable';
import Loader from 'react-spinners/HashLoader';

import { InterviewContext } from '../../../contexts/InterviewContext';
import InputString from 'chaire-lib-frontend/lib/components/input/InputString';
import { Link, useLocation, useParams } from 'react-router';
import { _booleish } from 'chaire-lib-common/lib/utils/LodashExtensions';

const loader = function Loading() {
    return <Loader size={30} color={'#aaaaaa'} loading={true} />;
};

const InterviewsComponent = Loadable({
    // TODO: move this to componentDidMount like in Monitoring.tsx
    loader: () => import('./InterviewSearchList'),
    loading: loader
});

const InterviewsByAccessCode: React.FC = () => {
    const location = useLocation();
    const params = useParams();
    const { t } = useTranslation('admin');

    const urlSearch = new URLSearchParams(location.search);
    const [currentCode, setCurrentCode] = React.useState(params.accessCode);
    const [createNewIfNoData] = React.useState(_booleish(urlSearch.get('autoCreate')) === true);
    const { state, dispatch } = React.useContext(InterviewContext);

    React.useEffect(() => {
        if (params.accessCode !== state.response.accessCode) {
            dispatch({ type: 'update_response', response: { accessCode: params.accessCode } });
        }
    }, [params.accessCode, state.response]);

    return (
        <div className="admin">
            <div className="survey-section__content apptr__form-container">
                <InputString
                    id="accessCodeSearchInput"
                    value={currentCode}
                    onValueUpdated={(newValue) => setCurrentCode(newValue.value)}
                ></InputString>
                <Link to={`/interviews/byCode/${currentCode}`}>
                    <button type="button" className={'survey-section__button button blue small'}>
                        {t('admin:interviewSearch:SearchByCode')}
                    </button>
                </Link>
            </div>
            <InterviewsComponent autoCreateIfNoData={createNewIfNoData} queryData={urlSearch} />
        </div>
    );
};

export default InterviewsByAccessCode;
