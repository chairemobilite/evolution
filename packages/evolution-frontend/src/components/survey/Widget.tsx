/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { withTranslation, WithTranslation } from 'react-i18next';
import React from 'react';
import _get from 'lodash/get';

import * as surveyHelper from 'evolution-common/lib/utils/helpers';
import Text from '../survey/Text';
import InfoMap from '../survey/InfoMap';
import Button from '../survey/Button';
import Question from '../survey/Question';
import { Group } from '../survey/GroupWidgets';
import { withSurveyContext, WithSurveyContextProps } from '../hoc/WithSurveyContextHoc';
import { UserFrontendInterviewAttributes, WidgetStatus } from '../../services/interviews/interview';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';
import { InterviewUpdateCallbacks } from 'evolution-common/lib/utils/helpers';
import { WidgetConfig } from 'evolution-common/lib/services/widgets';

export type WidgetProps = {
    currentWidgetShortname: string;
    nextWidgetShortname?: string;
    sectionName: string;
    interview: UserFrontendInterviewAttributes;
    errors?: { [path: string]: string };
    user: CliUser;
    loadingState: number;
};

export type InGroupWidgetProps = WidgetProps & {
    /** The path where to get the widget status for this widget. Defaults to `widgets`, but can be other in case the widget is for a group */
    widgetStatusPath: string;
    /** Path to prepend to the widget path, for example in groups */
    pathPrefix: string;
    groupedObjectId: string;
    parentObjectIds: { [shortname: string]: string };
};

type SingleWidgetProps = WidgetProps & {
    path?: string;
    customPath?: string;
    widgetStatus: WidgetStatus;
    groupedObjectId: string;
    parentObjectIds: { [shortname: string]: string };
};

// BaseWidget is a wrapper around the SingleWidget component that is responsible
// for setting the default values of some fields when the widget is not part of
// a group
const BaseWidget: React.FC<WidgetProps & WithTranslation & InterviewUpdateCallbacks & WithSurveyContextProps> = (
    props: WidgetProps & WithTranslation & InterviewUpdateCallbacks & WithSurveyContextProps
) => {
    const widgetConfig = props.surveyContext.widgets[props.currentWidgetShortname] as WidgetConfig;
    const widgetStatus = _get(props.interview, `widgets.${props.currentWidgetShortname}`, {}) as WidgetStatus;
    return (
        <SingleWidget
            {...props}
            path={widgetConfig?.path}
            customPath={(widgetConfig as any)?.customPath}
            widgetStatus={widgetStatus}
            groupedObjectId=""
            parentObjectIds={{}}
        />
    );
};

// BaseInGroupWidget is a wrapper around the SingleWidget component that is
// responsible for setting the default values of some fields when the widget is
// part of a group
const BaseInGroupWidget: React.FC<
    InGroupWidgetProps & WithTranslation & InterviewUpdateCallbacks & WithSurveyContextProps
> = (props: InGroupWidgetProps & WithTranslation & InterviewUpdateCallbacks & WithSurveyContextProps) => {
    const widgetConfig = props.surveyContext.widgets[props.currentWidgetShortname] as WidgetConfig;
    const widgetStatus = _get(
        props.interview,
        `${props.widgetStatusPath}.${props.currentWidgetShortname}`,
        {}
    ) as WidgetStatus;
    // Get the paths with prefix if it exists
    const widgetPath = props.pathPrefix && widgetConfig?.path ? `${props.pathPrefix}.${widgetConfig.path}` : undefined;
    const widgetCustomPath =
        props.pathPrefix && (widgetConfig as any)?.customPath
            ? `${props.pathPrefix}.${(widgetConfig as any).customPath}`
            : undefined;
    return (
        <SingleWidget
            {...props}
            path={widgetPath}
            customPath={widgetCustomPath}
            widgetStatus={widgetStatus}
            groupedObjectId={props.groupedObjectId}
            parentObjectIds={props.parentObjectIds}
        />
    );
};

/**
 * This is a single widget with all necessary properties to render it.
 */
const BaseSingleWidget: React.FC<
    SingleWidgetProps & WithTranslation & InterviewUpdateCallbacks & WithSurveyContextProps
> = (props: SingleWidgetProps & WithTranslation & InterviewUpdateCallbacks & WithSurveyContextProps) => {
    const widgetShortname = props.currentWidgetShortname;
    surveyHelper.devLog('%c rendering widget ' + widgetShortname, 'background: rgba(0,255,0,0.1);');
    const widgetConfig = props.surveyContext.widgets[widgetShortname] as WidgetConfig;
    if (widgetConfig === undefined) {
        console.error(`Widget is undefined: ${widgetShortname}`);
        if (props.surveyContext.devMode) {
            return (
                <div className="apptr__form-container two-columns question-invalid">{`Widget is undefined: ${widgetShortname}`}</div>
            );
        }
        return null;
    }

    const path = props.path
        ? surveyHelper.interpolatePath(props.interview, props.path)
        : `${props.sectionName}.${widgetShortname}`;
    const customPath = props.customPath ? surveyHelper.interpolatePath(props.interview, props.customPath) : undefined;
    // FIXME This should be typed and correctly got from the interview
    // FIXME 2 This component should be responsible to check the isVisible status and return null instead of each widget individually
    const widgetStatus = props.widgetStatus;
    const [isServerValid, serverErrorMessage] =
        props.errors && props.errors[path] ? [false, props.errors[path]] : [true, undefined];
    // Overwrite widget status with server validation data if invalid
    if (!isServerValid) {
        widgetStatus.isValid = false;
        widgetStatus.errorMessage = widgetStatus.errorMessage || serverErrorMessage;
    }

    const widgetProps: any = {
        // TODO: type default props
        path: path,
        customPath: customPath,
        key: path,
        shortname: widgetShortname,
        loadingState: props.loadingState,
        groupedObjectId: props.groupedObjectId,
        widgetConfig: widgetConfig,
        widgetStatus: widgetStatus,
        section: props.sectionName,
        interview: props.interview,
        user: props.user,
        startUpdateInterview: props.startUpdateInterview,
        startAddGroupedObjects: props.startAddGroupedObjects,
        startRemoveGroupedObjects: props.startRemoveGroupedObjects
    };

    switch (widgetConfig.type) {
    case 'text':
        return <Text {...widgetProps} />;
    case 'infoMap':
        return <InfoMap {...widgetProps} />;
    case 'button':
        return <Button {...widgetProps} />;
    case 'question': {
        // check for joined widgets:
        const nextWidgetConfig = props.nextWidgetShortname
            ? props.surveyContext.widgets[props.nextWidgetShortname]
            : undefined;
        const nextWidgetStatus = nextWidgetConfig
            ? (_get(props.interview, `widgets.${props.nextWidgetShortname}`, {}) as any)
            : undefined;
        const join =
                nextWidgetStatus &&
                ((nextWidgetConfig.joinWith === widgetShortname && nextWidgetStatus.isVisible) ||
                    (widgetConfig.joinWith === props.nextWidgetShortname && nextWidgetStatus.isVisible));
        widgetProps.join = join;
        return <Question {...widgetProps} />;
    }
    case 'group':
        return (
            <Group
                {...widgetProps}
                parentObjectIds={props.parentObjectIds}
                shortname={widgetShortname}
                errors={props.errors}
            />
        );
    }
};

export const Widget = withTranslation()(withSurveyContext(BaseWidget));

export const InGroupWidget = withTranslation()(withSurveyContext(BaseInGroupWidget));

const SingleWidget = withTranslation()(withSurveyContext(BaseSingleWidget));
