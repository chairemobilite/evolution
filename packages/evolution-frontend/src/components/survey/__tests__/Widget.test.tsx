/*
 * Copyright 2023, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import each from 'jest-each';

import { interviewAttributes } from '../../inputs/__tests__/interviewData.test';
import { Widget, InGroupWidget } from '../Widget';
import { WidgetStatus } from '../../../services/interviews/interview';

// Mock react-markdown and remark-gfm as they use syntax not supported by jest
jest.mock('react-markdown', () => 'Markdown');
jest.mock('remark-gfm', () => 'remark-gfm');

// Mock the react-datepicker files to avoid jest compilation errors in imported files
jest.mock('react-datepicker/dist/react-datepicker.css', () => {});
// Mock the react-input-range files to avoid jest compilation errors in imported files
jest.mock('react-input-range/src/js/input-range/default-class-names', () => ({
    activeTrack: 'input-range__track input-range__track--active',
    disabledInputRange: 'input-range input-range--disabled',
    inputRange: 'input-range',
    labelContainer: 'input-range__label-container',
    maxLabel: 'input-range__label input-range__label--max',
    minLabel: 'input-range__label input-range__label--min',
    slider: 'input-range__slider',
    sliderContainer: 'input-range__slider-container',
    track: 'input-range__track input-range__track--background',
    valueLabel: 'input-range__label input-range__label--value',
}));
jest.mock('react-input-range/lib/css/index.css', () => {});

const testWidgetShortname = 'test';
const testNextWidgetShortname = 'testNext';
const commonWidgetConfig = {
    type: 'question' as const,
    twoColumns: true,
    path: 'myTestPath',
    containsHtml: true,
    label: {
        fr: 'Texte en franÃ§ais',
        en: 'English text'
    }
};
const mockedContext = { sections: {}, widgets: { [testWidgetShortname]: {}, [testNextWidgetShortname]: commonWidgetConfig }, devMode: false, dispath: jest.fn()};
jest.mock('../../hoc/WithSurveyContextHoc', () => ({
    withSurveyContext: (Component) => (props) => <Component {...props} surveyContext={mockedContext}/> 
}));

const mockComponent = (props, name) => (
    <div>
        Mocked {name} Widget: 
        path: {props.path}
        customPath: {props.customPath}
        section: {props.section}
        loadingState: {props.loadingState}
        widgetConfig: {JSON.stringify(props.widgetConfig)}
        widgetStatus: {JSON.stringify(props.widgetStatus)}
        join: {props.join === true ? 'true' : 'false'}
        {name === 'Group' && (
            <>
                shortname: {props.shortname}
                parentObjectIds: {JSON.stringify(props.parentObjectIds)}
                errors: {props.errors ? JSON.stringify(props.errors) : undefined}
            </>
        )}
    </div>
);

// Mock the Question component
jest.mock('../Question', () => jest.fn((props) => mockComponent(props, 'Question')));

// Mock the Text component
jest.mock('../Text', () => jest.fn((props) => mockComponent(props, 'Text')));

// Mock the Button component
jest.mock('../Button', () => jest.fn((props) => mockComponent(props, 'Button')));

// Mock the InfoMap component
jest.mock('../InfoMap', () => jest.fn((props) => mockComponent(props, 'InfoMap')));

// Mock the Group component
jest.mock('../GroupWidgets', () => ({
    Group: jest.fn((props) => mockComponent(props, 'Group'))
}));

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
};

const defaultWidgetStatus: WidgetStatus = {
    path: commonWidgetConfig.path,
    isVisible: true,
    modalIsOpen: false,
    isDisabled: false,
    isCollapsed: false,
    isEmpty: false,
    isCustomEmpty: false,
    isValid: true,
    isResponded: true,
    isCustomResponded: false,
    errorMessage: undefined,
    currentUpdateKey: 1,
    value: null
};

const groupWidgetStatus: WidgetStatus = {
    ..._cloneDeep(defaultWidgetStatus),
    path: 'myGroupPath'
};

const frontendInterviewAttributes = {
    ...interviewAttributes,
    previousWidgets: {},
    previousGroups: {},
    widgets: {
        [testWidgetShortname]: _cloneDeep(defaultWidgetStatus),
        [testNextWidgetShortname]: _cloneDeep(defaultWidgetStatus)
    },
    groups: {
        myGroup: {
            myGroupId: {
                [testWidgetShortname]: _cloneDeep(groupWidgetStatus),
                [testNextWidgetShortname]: _cloneDeep(groupWidgetStatus)
            }
        }
    },
    visibleWidgets: [],
    allWidgetsValid: true
};

each([
    ['Text', { type: 'text', text: 'test text' }],
    ['InfoMap', { type: 'infoMap', title: 'info map title', geojsons: () => {} }],
    ['Button', { type: 'button', action: jest.fn(), label: 'test button label' }],
    ['Question', { type: 'question', path: 'test', label: 'test question', inputType: 'string' }],
    ['Question with custom path', { type: 'question', path: 'test', customPath: 'test.custom', label: 'test question', inputType: 'string' }],
    ['Question with join', { type: 'question', path: 'test', label: 'test question', inputType: 'string', joinWith: testNextWidgetShortname }],
    ['Group', { type: 'group', path: 'test', widgets: [testNextWidgetShortname] }],
    ['undefined', undefined]
]).describe('Widget of type %s', (_widget, widgetConfig) => {

    test('Render widget', () => {
        // Set the widget config in the context. Has to be in the test, since
        // it's a shared variable and otherwise it's racy with other tests
        mockedContext.widgets[testWidgetShortname] = widgetConfig;

        // What to check in snapshots
        // Expected path: should be the one specified in the widget, otherwise prefixed with "Test."
        // Expected section: Test
        // Expected widget status: should be the one with path as "myTestPath"
        // For groups, the additional parameters should be present but empty

        const wrapper = TestRenderer.create(
            <Widget
                currentWidgetShortname={testWidgetShortname}
                nextWidgetShortname={testNextWidgetShortname}
                sectionName='Test'
                interview={frontendInterviewAttributes}
                loadingState={0}
                errors={{}}
                user={userAttributes}
                startUpdateInterview={jest.fn()}
                startAddGroupedObjects={jest.fn()}
                startRemoveGroupedObjects={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Render in group widget', () => {
        // Set the widget config in the context. Has to be in the test, since
        // it's a shared variable and otherwise it's racy with other tests
        mockedContext.widgets[testWidgetShortname] = widgetConfig;
        const groupedShortname = 'myGroupName';

        // What to check in snapshots
        // Expected path: should be prefixed with groupPath.uuid.groupName for widgets with a path specified, otherwise prefixed with "myGroupName."
        // Expected section: myGroupName
        // Expected widget status: should be the one with path as "myGroupPath"
        // For groups, the additional parameters should match what is sent here

        const wrapper = TestRenderer.create(
            <InGroupWidget
                currentWidgetShortname={testWidgetShortname}
                nextWidgetShortname={testNextWidgetShortname}
                sectionName={groupedShortname}
                interview={frontendInterviewAttributes}
                loadingState={0}
                errors={{}}
                user={userAttributes}
                startUpdateInterview={jest.fn()}
                startAddGroupedObjects={jest.fn()}
                startRemoveGroupedObjects={jest.fn()}
                widgetStatusPath='groups.myGroup.myGroupId'
                pathPrefix='groupPath.uuid.groupName'
                groupedObjectId='myGroupId'
                parentObjectIds={{[groupedShortname]: 'myGroupId'}}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

});

describe('With server errors', () => {

    test('Render Question with server error', () => {
        // In the snapshot, expect the widget status to have the extra server error in it
        const questionWidgetConfig = {
            type: 'question',
            path: 'test',
            label: 'test question',
            inputType: 'string'
        };
        mockedContext.widgets[testWidgetShortname] = questionWidgetConfig;

        const wrapper = TestRenderer.create(
            <Widget
                currentWidgetShortname={testWidgetShortname}
                nextWidgetShortname={testNextWidgetShortname}
                sectionName='Test'
                interview={frontendInterviewAttributes}
                loadingState={0}
                errors={{
                    [testWidgetShortname]: 'Server error message'
                }}
                user={userAttributes}
                startUpdateInterview={jest.fn()}
                startAddGroupedObjects={jest.fn()}
                startRemoveGroupedObjects={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Render Group with server error, shoud be passed along', () => {
        // In the snapshot, expect the widget status to have the extra server
        // error in it. The errors prop should also be passed along to the Group
        // widget
        const questionWidgetConfig = { type: 'group', path: 'test', widgets: [testNextWidgetShortname] };
        mockedContext.widgets[testWidgetShortname] = questionWidgetConfig;

        const wrapper = TestRenderer.create(
            <Widget
                currentWidgetShortname={testWidgetShortname}
                nextWidgetShortname={testNextWidgetShortname}
                sectionName='Test'
                interview={frontendInterviewAttributes}
                loadingState={0}
                errors={{
                    [testWidgetShortname]: 'Server error message'
                }}
                user={userAttributes}
                startUpdateInterview={jest.fn()}
                startAddGroupedObjects={jest.fn()}
                startRemoveGroupedObjects={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();

    })

});
