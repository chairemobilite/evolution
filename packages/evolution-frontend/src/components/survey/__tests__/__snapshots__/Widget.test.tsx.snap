// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget of type Button Render in group widget 1`] = `
<div>
  Mocked 
  Button
   Widget: path: 
  myGroupName.test
  customPath: 
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"button","label":"test button label"}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Button Render widget 1`] = `
<div>
  Mocked 
  Button
   Widget: path: 
  Test.test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"button","label":"test button label"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Group Render in group widget 1`] = `
<div>
  Mocked 
  Group
   Widget: path: 
  groupPath.uuid.groupName.test
  customPath: 
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"group","path":"test","widgets":["testNext"]}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
  shortname: 
  test
  parentObjectIds: 
  {"myGroupName":"myGroupId"}
  errors: 
  {}
</div>
`;

exports[`Widget of type Group Render widget 1`] = `
<div>
  Mocked 
  Group
   Widget: path: 
  test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"group","path":"test","widgets":["testNext"]}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
  shortname: 
  test
  parentObjectIds: 
  {}
  errors: 
  {}
</div>
`;

exports[`Widget of type InfoMap Render in group widget 1`] = `
<div>
  Mocked 
  InfoMap
   Widget: path: 
  myGroupName.test
  customPath: 
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"infoMap","title":"info map title"}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type InfoMap Render widget 1`] = `
<div>
  Mocked 
  InfoMap
   Widget: path: 
  Test.test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"infoMap","title":"info map title"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Question Render in group widget 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  groupPath.uuid.groupName.test
  customPath: 
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","label":"test question","inputType":"string"}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Question Render widget 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","label":"test question","inputType":"string"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Question with custom path Render in group widget 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  groupPath.uuid.groupName.test
  customPath: 
  groupPath.uuid.groupName.test.custom
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","customPath":"test.custom","label":"test question","inputType":"string"}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Question with custom path Render widget 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  test
  customPath: 
  test.custom
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","customPath":"test.custom","label":"test question","inputType":"string"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Question with join Render in group widget 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  groupPath.uuid.groupName.test
  customPath: 
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","label":"test question","inputType":"string","joinWith":"testNext"}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  true
</div>
`;

exports[`Widget of type Question with join Render widget 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","label":"test question","inputType":"string","joinWith":"testNext"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  true
</div>
`;

exports[`Widget of type Text Render in group widget 1`] = `
<div>
  Mocked 
  Text
   Widget: path: 
  myGroupName.test
  customPath: 
  section: 
  myGroupName
  loadingState: 
  0
  widgetConfig: 
  {"type":"text","text":"test text"}
  widgetStatus: 
  {"path":"myGroupPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type Text Render widget 1`] = `
<div>
  Mocked 
  Text
   Widget: path: 
  Test.test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"text","text":"test text"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":true,"isResponded":true,"isCustomResponded":false,"currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;

exports[`Widget of type undefined Render in group widget 1`] = `null`;

exports[`Widget of type undefined Render widget 1`] = `null`;

exports[`With server errors Render Group with server error, shoud be passed along 1`] = `
<div>
  Mocked 
  Group
   Widget: path: 
  test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"group","path":"test","widgets":["testNext"]}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":false,"isResponded":true,"isCustomResponded":false,"errorMessage":"Server error message","currentUpdateKey":1,"value":null}
  join: 
  false
  shortname: 
  test
  parentObjectIds: 
  {}
  errors: 
  {"test":"Server error message"}
</div>
`;

exports[`With server errors Render Question with server error 1`] = `
<div>
  Mocked 
  Question
   Widget: path: 
  test
  customPath: 
  section: 
  Test
  loadingState: 
  0
  widgetConfig: 
  {"type":"question","path":"test","label":"test question","inputType":"string"}
  widgetStatus: 
  {"path":"myTestPath","isVisible":true,"modalIsOpen":false,"isDisabled":false,"isCollapsed":false,"isEmpty":false,"isCustomEmpty":false,"isValid":false,"isResponded":true,"isCustomResponded":false,"errorMessage":"Server error message","currentUpdateKey":1,"value":null}
  join: 
  false
</div>
`;
